### PR TITLE
Add odin.password.hash view

### PR DESCRIPTION
**Pull request by tleekkul**

### CODE REVIEW

File: Cpp/odin-views/password.cpp [hunk 1]
```diff
@@ -185,5 +186,45 @@ namespace {
...
+            fostlib::json body = fostlib::json::parse(body_str);
+            auto hash_value = fostgres::datum(
+                    config["hash"], std::vector<fostlib::string>{}, body, req);
+            auto verify = fostgres::datum(
+                    config["verify"], std::vector<fostlib::string>{}, body,
+                    req);
+            if (not hash_value || not verify || hash_value != verify) {
+                return respond("Hashing failed", 422);
+            }
+            auto const hash_result = odin::hash_password(
```
Comment (tleekkul) [inline Cpp/odin-views/password.cpp]: How to test this line? @KayEss

Comment (KayEss) [reply Cpp/odin-views/password.cpp]: Can you can pass headers into the test through the request constructor. You can get response headers from the response object. Either of those help? Look at the [`request` and `response` objects](https://github.com/KayEss/fost-internet/blob/master/Cpp/include/fost/http.server.hpp)

File: Cpp/odin-views/password.cpp [hunk 2]
```diff
@@ -185,5 +186,45 @@ namespace {
...
+        std::pair<boost::shared_ptr<fostlib::mime>, int> operator()(
+                const fostlib::json &config,
+                const fostlib::string &path,
+                fostlib::http::server::request &req,
+                const fostlib::host &host) const {
+            if (!config.has_key("hash") || !config.has_key("verify")
+                || !config.has_key("then")) {
+                throw fostlib::exceptions::not_implemented(
+                        __PRETTY_FUNCTION__,
+                        "Must supply both hash, verify and then in the "
```
Comment (KayEss) [inline Cpp/odin-views/password.cpp]: I couldn't parse that text to start with. Maybe put single quotes around the key words?

    "Must supply 'hash', 'verify' and 'then' in the "

Comment (KayEss) [reply Cpp/odin-views/password.cpp]: Although `verify` is presumably optional, but highly recommended?

File: Cpp/odin-views/password.cpp [hunk 3]
```diff
@@ -185,5 +186,45 @@ namespace {
...
+                        "configuration");
+            }
+
+            auto body_str = fostlib::coerce<fostlib::string>(
+                    fostlib::coerce<fostlib::utf8_string>(req.data()->data()));
+            fostlib::json body = fostlib::json::parse(body_str);
+            auto hash_value = fostgres::datum(
+                    config["hash"], std::vector<fostlib::string>{}, body, req);
+            auto verify = fostgres::datum(
+                    config["verify"], std::vector<fostlib::string>{}, body,
```
Comment (KayEss) [inline Cpp/odin-views/password.cpp]: I wonder if you can just pass `{}` for the vector and not have to spell out the full type.

File: Cpp/odin-views/password.cpp [hunk 4]
```diff
@@ -185,5 +186,45 @@ namespace {
...
+    const class password_hash : public fostlib::urlhandler::view {
+      public:
+        password_hash() : view("odin.password.hash") {}
+
+        std::pair<boost::shared_ptr<fostlib::mime>, int> operator()(
+                const fostlib::json &config,
+                const fostlib::string &path,
+                fostlib::http::server::request &req,
+                const fostlib::host &host) const {
+            if (!config.has_key("hash") || !config.has_key("verify")
```
Comment (KayEss) [inline Cpp/odin-views/password.cpp]: Should probably make sure that the method is `POST` too shouldn't it?

Comment (tleekkul) [reply Cpp/odin-views/password.cpp]: I think we should support every method that has a request body (POST, PUT, PATCH)

File: Cpp/odin-views/password.cpp (hunk 2 above)
Comment (tleekkul) [reply Cpp/odin-views/password.cpp]: for now, I think it's ok if you set the config to be the same as `hash` key.
e.g.```
{"hash": ["body", "password"], "verify": ["body", "password"]}
```

File: Cpp/odin-views/password.cpp (hunk 1 above)
Comment (KayEss) [reply Cpp/odin-views/password.cpp]: Yeah

File: Cpp/odin-views/password.cpp (hunk 4 above)
Comment (KayEss) [reply Cpp/odin-views/password.cpp]: Yeah

File: Cpp/odin-views/password.cpp (hunk 1 above)
Comment (tleekkul) [reply Cpp/odin-views/password.cpp]: No, I think those are not help.

File: Cpp/odin-views/password.hash.tests.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                const fostlib::host &host) const {
+            FSL_CHECK_EQ(req.headers().exists("__hash"), true);
+            const auto hash = req.headers()["__hash"].value();
+
+            FSL_CHECK_EQ(req.headers().exists("__hash_process"), true);
+            const auto hash_process = fostlib::json::parse(
+                    req.headers()["__hash_process"].value());
+            FSL_CHECK_EQ(hash_process["length"], fostlib::json{32});
+            FSL_CHECK_EQ(hash_process["name"], fostlib::json{"pbkdf2-sha256"});
+            FSL_CHECK_EQ(hash_process["rounds"], fostlib::json{300000});
```
Comment (KayEss) [inline Cpp/odin-views/password.hash.tests.cpp]: This will fail the moment we change this default, but I think it's OK for now.

**Pull request merged**